### PR TITLE
Role and Privilege Controllers

### PR DESCRIPTION
**Pull request by EricRBedard**
Creating Controllers for Privilege and Role

### CODE REVIEW

File: src/main/java/sysc4806/graduateAdmissions/controller/PrivilegeController.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    @GetMapping("")
+    public ResponseEntity getPrivilege(@RequestParam(required=false) Long id) {
+        if(id == null)
+            return ResponseEntity.status(HttpStatus.OK).body(repository.findAll());
+        else
+            return ResponseEntity.status(HttpStatus.OK).body(repository.findById(id));
+    }
+
+    /**
+     * get all Privilege of a specified Owner
```
Comment (LukeANewton) [inline src/main/java/sysc4806/graduateAdmissions/controller/PrivilegeController.java]: the way you have been commenting methods dealing with Owners makes it seem like an Owner is like a User, in that it describes a specific person. The Owner enum does not do this, before I approve, I want to make sure you know this

File: src/main/java/sysc4806/graduateAdmissions/repositories/PrivilegeRepository.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.List;
+
+/**
+ * This repository will be used for persistent storage of Privileges
+ *
+ * @author Eric
+ */
+public interface PrivilegeRepository extends CrudRepository<Privilege, Long> {
+    List<Privilege> findByOwner(Owner owner);
```
Comment (LukeANewton) [inline src/main/java/sysc4806/graduateAdmissions/repositories/PrivilegeRepository.java]: Why allow Privileges to be found by only Owner, and not by Operation or Target?

File: src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    /**
+     * remove Privilege from Role
+     *
+     * @param id id of the Role to be updated
+     * @param privilege the privilege to be removed
+     * @return ResponseEntity describing the outcome of the operation
+     */
+    @PostMapping("remove")
+    public ResponseEntity removeRolePrivilege(@RequestParam Long id, @RequestBody Privilege privilege){
+        Optional<Role> role = repository.findById(id);
```
Comment (LukeANewton) [inline src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java]: does this work? The Role object does not have a long id, the id is the name (a string)

File: src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * create a new Role
+     *
+     * @param role the Role to be added to the system
+     * @return ResponseEntity describing the outcome of the operation
+     */
+    @ResponseBody
+    @PostMapping("create")
+    public ResponseEntity createRoleForm(@RequestBody() Role role) {
```
Comment (LukeANewton) [inline src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java]: nothing about this suggests that there is a from involved, besides the name of the method

File: src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java (hunk 3 above)
Comment (EricRBedard) [reply src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java]: you're right, autocomplete betrayed me. will change this (and the other places i used findById())

File: src/main/java/sysc4806/graduateAdmissions/controller/PrivilegeController.java (hunk 1 above)
Comment (EricRBedard) [reply src/main/java/sysc4806/graduateAdmissions/controller/PrivilegeController.java]: didn't mean it that way, ill try rewording the comments

File: src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java (hunk 4 above)
Comment (EricRBedard) [reply src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java]: missed this when changing controller to return JSON, will fix

File: src/main/java/sysc4806/graduateAdmissions/repositories/PrivilegeRepository.java (hunk 2 above)
Comment (EricRBedard) [reply src/main/java/sysc4806/graduateAdmissions/repositories/PrivilegeRepository.java]: Good point, added those and controller methods for them

File: src/test/java/sysc4806/graduateAdmissions/controller/RoleControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    /* Test update Role */
+    @Test
+    public void testUpdateRole() throws Exception {
+        mockMvc.perform(post("/role/update").contentType(APPLICATION_JSON_UTF8)
+                .content(toJson(
+                        Role.builder().roleName("Student").build())))
+                .andExpect(status().isOk());
+    }
```
Comment (CameronRushton) [inline src/test/java/sysc4806/graduateAdmissions/controller/RoleControllerTest.java]: What is the difference between this update role and the create role? (I actually did mine wrong in my PR too). The controller has two of the same methods but different endpoint and return string, but does the exact same thing.
[This article did a good job of explaining how the code should look](http://progressivecoder.com/rest-api-using-spring-boot-insert-update/) (minus the service layer what we have yet to implement.)
TL;DR:
|                 | PostMapping | PutMapping                                                     |   |   |
|-----------------|-------------|----------------------------------------------------------------|---|---|
| Returns         | 201 CREATED | 200 OK                                                         |   |   |
| Takes path var? | No          | Yes (xx/{id})                                                  |   |   |
| Code on Error   | 5xx series  | 404 when xx.id<br>not found.<br>(use existsById)<br>5xx series |   |   |

**Comment by CameronRushton** [comment]
I had my review open way too long it seems haha

File: src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    /**
+     * remove Privilege from Role
+     *
+     * @param name roleName of the Role to be updated
+     * @param privilege the privilege to be removed
+     * @return ResponseEntity describing the outcome of the operation
+     */
+    @PostMapping("remove")
+    public ResponseEntity removeRolePrivilege(@RequestParam String name, @RequestBody Privilege privilege){
```
Comment (CameronRushton) [inline src/main/java/sysc4806/graduateAdmissions/controller/RoleController.java]: This is a POST when it should be DELETE
